pkg/util/limiter/per_query_limiter.go

```diff
@@ -0,0 +1,119 @@
+package limiter
+
+import (
+	"context"
+	"fmt"
+	"github.com/cortexproject/cortex/pkg/cortexpb"
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/validation"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"go.uber.org/atomic"
+	"sync"
+)
+
+type PerQueryLimiter struct {
+	uniqueSeriesMx sync.RWMutex
+	uniqueSeries   map[model.Fingerprint]struct{}
+
+	chunkBytesCount       *atomic.Int32
+	maxSeriesPerQuery     int
+	maxChunkBytesPerQuery int
+}
+type perQueryLimiterCtxMarker struct{}
+
+var (
+	pqlCtxKey           = &perQueryLimiterCtxMarker{}
+	errMaxSeriesHit     = "The query hit the max number of series limit while fetching chunks %s (limit: %d)"
+	errMaxChunkBytesHit = "The query hit the max number of chunk bytes limit while fetching chunks %s (limit: %d)"
+)
+
+// NewPerQueryLimiter makes a new per-query rate limiter. Each per-query limiter
+// is configured using the `maxSeriesPerQuery` and `maxChunkBytesPerQuery` limits.
+func NewPerQueryLimiter(maxSeriesPerQuery int, maxChunkBytesPerQuery int) *PerQueryLimiter {
+	return &PerQueryLimiter{
+		uniqueSeriesMx: sync.RWMutex{},
+		uniqueSeries:   map[model.Fingerprint]struct{}{},
+
+		chunkBytesCount: atomic.NewInt32(0),
+
+		maxChunkBytesPerQuery: maxChunkBytesPerQuery,
+		maxSeriesPerQuery:     maxSeriesPerQuery,
+	}
+}
+
+func NewPerQueryLimiterOnContext(ctx context.Context, maxSeriesPerQuery int, maxChunkBytesPerQuery int) context.Context {
```

Comment:
[nit] Do we need this function at all? The logic can be achieve concatenating the `AddPerQueryLimiterToContext()` and `NewPerQueryLimiter()` together.